wordpress-mobile/WordPress-Android
# [Jetpack Focus] Revamp Landing Screen - Sensor animation

Fixes #17075

With this PR we're using accelerometer sensor data (y-axis acceleration) to influence the speed of the vertical scrolling text from the background of the Jetpack App Landing Screen.

~~⚠️ **Important Notice** This PR should target `trunk` after `feature/revamp-landing-screen` is merged.
PR #17203 should be merged first, then we can target `trunk` with this PR and set it `Ready for review`.~~

To test:

<details>
<summary><b>Prerequisite</b> - Enable the <code>LandingScreenRevamp</code> feature flag</summary>

- Either `login` → `Me` → `App Settings` → `Debug Settings` → enable `LandingScreenRevamp` → `RESTART THE APP` → `logout` or
- Apply this hard-coded solution:

<details>
<summary>Patch to enable the <code>LANDING_SCREEN_REVAMP</code> build flag</summary>

```diff
diff --git a/WordPress/build.gradle b/WordPress/build.gradle
index 63260a4ca8..fe999fa92f 100644
--- a/WordPress/build.gradle
+++ b/WordPress/build.gradle
@@ -112,7 +112,7 @@ android {
         buildConfigField "boolean", "READER_COMMENTS_MODERATION", "false"
         buildConfigField "boolean", "SITE_INTENT_QUESTION", "true"
         buildConfigField "boolean", "SITE_NAME", "false"
-        buildConfigField "boolean", "LANDING_SCREEN_REVAMP", "false"
+        buildConfigField "boolean", "LANDING_SCREEN_REVAMP", "true"
         buildConfigField "boolean", "LAND_ON_THE_EDITOR", "false"
         buildConfigField "boolean", "BLOGGING_PROMPTS", "false"
         buildConfigField "boolean", "QUICK_START_EXISTING_USERS_V2", "false"
```  
</details>

---

</details>

##### 1️⃣ Test Case 1: Happy Flow

0. Open the JP app and make sure you're logged out
1. Make sure you're holding the device in your hand slightly rotated vertically, as illustrated:
   <img width="100" src="https://user-images.githubusercontent.com/4588074/191999204-611285aa-9bc0-4cdc-8384-bed2ef6f07ff.png">
3. **Expect** the large text on the new login screen to scroll down by default at an accelerated speed
4. Lay the device flat on your office table
5. **Expect** The text to scroll **down** at a slower speed
6. Pick up the device and rotate it slightly inverse than in the illustration (top of the phone should be below its bottom)
7. **Expect** The text to scroll **up**

##### 2️⃣ Test Case 2: Feature flag off

0. Disable the <code>LandingScreenRevamp</code> feature flag (revert change from **Prerequisite**)
1. Open the JP app and make sure you're logged out
2. **Expect** To be presented with the previous design
3. Tap any of the 2 buttons
4. **Expect** the Login flows to still work as usual.

##### 3️⃣ Test Case 3: Sensors off

1. Enable developer mode:
   `Settings` → `About phone / emulated device` → tap `Build number` 7 times
2. Enable `Sensors Off quick setting:
   `Go to `Settings` → `System` → `Developer options` → `Quick settings developer tiles` tap `Sensors Off`
3. Toggle Sensors off:
   Swipe down from the top of the screen and tap `Sensors Off`
4. Open the JP app and make sure you're logged out
5. **Expect** the large text to scroll down at a speed similar to the normal position speed when sensors are enabled
   (step 3 of Test Case 1)

##### 4️⃣ Test Case 4: No accelerometer sensor

1. Apply this patch to simulate missing accelerometer sensor on the device:

```diff
diff --git a/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt b/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt
index 99e8570448..c57f1aaf1e 100644
--- a/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt
+++ b/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt
@@ -62,7 +62,7 @@ class LoginPrologueRevampedViewModel @Inject constructor(
 
         override fun onActive() {
             super.onActive()
-            val sensor = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
+            val sensor: Sensor? = null
             sensorManager.registerListener(this, sensor, SensorManager.SENSOR_DELAY_GAME)
         }
```

2. Open the JP app and make sure you're logged out

3. **Expect** the large text to scroll down at a speed similar to the normal position speed when sensors are enabled

<details>
<summary><h4>🎬 Video Preview</h4></summary>

https://user-images.githubusercontent.com/4588074/192004342-b2955e14-95b0-45d3-87ff-9db4f5ccce28.mp4

- Initially the device is rotated to `-45°` (reflecting a normal use position)
- At second `5` the device is rotated to `-90°` (reflecting flat on table position)
- At second `13` the device is rotated to `-105°` (inverted rotation that triggers reverse scrolling)

---

</details>

## Regression Notes
4. Potential unintended areas of impact
   New Jetpack Landing Screen when sensor is not available.

8. What I did to test those areas of impact (or what existing automated tests I relied on)
   Tested manually different configs: sensors disabled / missing and added test cases for these.

9. What automated tests I added (or what prevented me from doing so)
   None, we're interacting with platform code and UI animation code.

PR submission checklist:
